File: internal/model/amp.go
package model

import "time"

// WebSearchMode constants
const (
	WebSearchModeUpstream     = "upstream"      // 上游代理（不做修改）
	WebSearchModeBuiltinFree  = "builtin_free"  // 内置免费搜索（强制 isFreeTierRequest=true）
	WebSearchModeLocalDDG     = "local_duckduckgo" // 本地 DuckDuckGo 搜索
)

type AmpSettings struct {
	ID                 string    `json:"id"`
	UserID             string    `json:"user_id"`
	UpstreamURL        string    `json:"upstream_url"`
	UpstreamAPIKey     string    `json:"-"`
	ModelMappingsJSON  string    `json:"-"`
	ForceModelMappings bool      `json:"force_model_mappings"`
	Enabled            bool      `json:"enabled"`
	WebSearchMode      string    `json:"web_search_mode"` // upstream | builtin_free | local_duckduckgo
	NativeMode         bool      `json:"native_mode"`
	ShowBalanceInAd    bool      `json:"show_balance_in_ad"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type ModelMapping struct {
	From            string   `json:"from"`
	To              string   `json:"to"`
	Regex           bool     `json:"regex"`
	ThinkingLevel   string   `json:"thinkingLevel,omitempty"`
	PseudoNonStream bool     `json:"pseudoNonStream,omitempty"`
	AuditKeywords   []string `json:"auditKeywords,omitempty"`
}

type UserAPIKey struct {
	ID         string     `json:"id"`
	UserID     string     `json:"user_id"`
	Name       string     `json:"name"`
	KeyHash    string     `json:"-"`
	APIKey     string     `json:"-"`
	Prefix     string     `json:"prefix"`
	LastUsedAt *time.Time `json:"last_used_at,omitempty"`
	LastUsed   *time.Time `json:"last_used,omitempty"`
	ExpiresAt  *time.Time `json:"expires_at,omitempty"`
	RevokedAt  *time.Time `json:"revoked_at,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
}

// Request/Response 结构体

type AmpSettingsRequest struct {
	UpstreamURL        string         `json:"upstreamUrl"`
	UpstreamAPIKey     string         `json:"upstreamApiKey,omitempty"`
	ModelMappings      []ModelMapping `json:"modelMappings,omitempty"`
	ForceModelMappings bool           `json:"forceModelMappings"`
	Enabled            bool           `json:"enabled"`
	WebSearchMode      string         `json:"webSearchMode,omitempty"` // upstream | builtin_free | local_duckduckgo
	NativeMode         bool           `json:"nativeMode"`
	ShowBalanceInAd    *bool          `json:"showBalanceInAd,omitempty"`
}

type AmpSettingsResponse struct {
	UpstreamURL        string         `json:"upstreamUrl"`
	ModelMappings      []ModelMapping `json:"modelMappings"`
	ForceModelMappings bool           `json:"forceModelMappings"`
	Enabled            bool           `json:"enabled"`
	HasAPIKey          bool           `json:"apiKeySet"`
	WebSearchMode      string         `json:"webSearchMode"` // upstream | builtin_free | local_duckduckgo
	NativeMode         bool           `json:"nativeMode"`
	ShowBalanceInAd    bool           `json:"showBalanceInAd"`
	CreatedAt          time.Time      `json:"createdAt,omitempty"`
	UpdatedAt          time.Time      `json:"updatedAt,omitempty"`
}

type TestConnectionResponse struct {
	Success   bool   `json:"success"`
	Message   string `json:"message"`
	LatencyMs int64  `json:"latencyMs,omitempty"`
}

type CreateAPIKeyRequest struct {
	Name string `json:"name" binding:"required,min=1,max=64"`
}

type CreateAPIKeyResponse struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Prefix    string    `json:"prefix"`
	APIKey    string    `json:"apiKey"`
	CreatedAt time.Time `json:"createdAt"`
	Message   string    `json:"message"`
}

type APIKeyRevealResponse struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Prefix    string    `json:"prefix"`
	APIKey    string    `json:"apiKey"`
	CreatedAt time.Time `json:"createdAt"`
}

type APIKeyListItem struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	Prefix    string     `json:"prefix"`
	CreatedAt time.Time  `json:"createdAt"`
	RevokedAt *time.Time `json:"revokedAt,omitempty"`
	LastUsed  *time.Time `json:"lastUsedAt,omitempty"`
	IsActive  bool       `json:"isActive"`
}

type BootstrapResponse struct {
	HasSettings bool `json:"hasSettings"`
	HasAPIKey   bool `json:"hasApiKey"`
}

// RequestLogStatus 请求日志状态
type RequestLogStatus string

const (
	RequestLogStatusPending RequestLogStatus = "pending"
	RequestLogStatusSuccess RequestLogStatus = "success"
	RequestLogStatusError   RequestLogStatus = "error"
)

// RequestLog 请求日志记录
type RequestLog struct {
	ID                       string           `json:"id"`
	CreatedAt                string           `json:"createdAt"`
	UpdatedAt                *string          `json:"updatedAt,omitempty"`
	Status                   RequestLogStatus `json:"status"`
	UserID                   string           `json:"userId"`
	Username                 *string          `json:"username,omitempty"`
	APIKeyID                 string           `json:"apiKeyId"`
	OriginalModel            *string          `json:"originalModel,omitempty"`
	MappedModel              *string          `json:"mappedModel,omitempty"`
	Provider                 *string          `json:"provider,omitempty"`
	ChannelID                *string          `json:"channelId,omitempty"`
	Endpoint                 *string          `json:"endpoint,omitempty"`
	Method                   string           `json:"method"`
	Path                     string           `json:"path"`
	StatusCode               int              `json:"statusCode"`
	LatencyMs                int64            `json:"latencyMs"`
	IsStreaming              bool             `json:"isStreaming"`
	InputTokens              *int             `json:"inputTokens,omitempty"`
	OutputTokens             *int             `json:"outputTokens,omitempty"`
	CacheReadInputTokens     *int             `json:"cacheReadInputTokens,omitempty"`
	CacheCreationInputTokens *int             `json:"cacheCreationInputTokens,omitempty"`
	ErrorType                *string          `json:"errorType,omitempty"`
	RequestID                *string          `json:"requestId,omitempty"`
	ThinkingLevel            *string          `json:"thinkingLevel,omitempty"` // 思维等级
	OutputPreview            *string          `json:"outputPreview,omitempty"` // 响应输出预览（前200字符）
	// 成本相关字段
	CostMicros   *int64  `json:"costMicros,omitempty"`   // 成本（微美元，USD * 1e6）
	CostUsd      *string `json:"costUsd,omitempty"`      // 成本（USD，用于展示）
	PricingModel *string `json:"pricingModel,omitempty"` // 计价模型名
}

// RequestLogListResponse 请求日志列表响应
type RequestLogListResponse struct {
	Items    []RequestLog `json:"items"`
	Total    int64        `json:"total"`
	Page     int          `json:"page"`
	PageSize int          `json:"pageSize"`
}

// UsageSummary 用量统计
type UsageSummary struct {
	GroupKey                    string `json:"groupKey"`
	InputTokensSum              int64  `json:"inputTokensSum"`
	OutputTokensSum             int64  `json:"outputTokensSum"`
	CacheReadInputTokensSum     int64  `json:"cacheReadInputTokensSum"`
	CacheCreationInputTokensSum int64  `json:"cacheCreationInputTokensSum"`
	RequestCount                int64  `json:"requestCount"`
	ErrorCount                  int64  `json:"errorCount"`
	CostMicrosSum               int64  `json:"costMicrosSum"`  // 总成本（微美元）
	CostUsdSum                  string `json:"costUsdSum"`     // 总成本（USD 展示）
}

// UsageSummaryResponse 用量统计响应
type UsageSummaryResponse struct {
	Items []UsageSummary `json:"items"`
}

// RequestLogDetail 请求日志详情（包含请求/响应头和体）
type RequestLogDetail struct {
	RequestID              string            `json:"requestId"`
	RequestHeaders         map[string]string `json:"requestHeaders"`
	RequestBody            string            `json:"requestBody"`
	TranslatedRequestBody  string            `json:"translatedRequestBody,omitempty"`  // 翻译后发送给上游的请求
	ResponseHeaders        map[string]string `json:"responseHeaders"`
	ResponseBody           string            `json:"responseBody"`
	TranslatedResponseBody string            `json:"translatedResponseBody,omitempty"` // 翻译后发送给客户端的响应
	CreatedAt              time.Time         `json:"createdAt"`
}

File: web/src/api/amp.ts
import { authFetch } from './client'

const API_BASE = '/api/me/amp'

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: '请求失败' }))
    throw new Error(error.error || '请求失败')
  }
  return response.json()
}

// Types
export type WebSearchMode = 'upstream' | 'builtin_free' | 'local_duckduckgo'

export interface ModelMapping {
  from: string
  to: string
  regex: boolean
  thinkingLevel?: 'low' | 'medium' | 'high' | 'xhigh' | ''
  pseudoNonStream?: boolean
  auditKeywords?: string[]
  ampOnly?: boolean
  fastMode?: boolean
}

export interface AmpSettings {
  upstreamUrl: string
  apiKeySet: boolean
  modelMappings: ModelMapping[]
  enabled: boolean
  nativeMode: boolean
  webSearchMode?: WebSearchMode
  showBalanceInAd?: boolean
  socks5ProxySet?: boolean
}

export interface UpdateAmpSettingsRequest {
  upstreamUrl?: string
  upstreamApiKey?: string
  modelMappings?: ModelMapping[]
  enabled?: boolean
  nativeMode?: boolean
  webSearchMode?: WebSearchMode
  showBalanceInAd?: boolean
  socks5Proxy?: string
}

export interface TestResult {
  success: boolean
  message: string
  latencyMs?: number
}

export interface APIKey {
  id: string
  name: string
  prefix: string
  lastUsedAt: string | null
  createdAt: string
}

export interface CreateAPIKeyResponse {
  id: string
  name: string
  prefix: string
  apiKey: string
  createdAt: string
  message: string
}

export interface APIKeyRevealResponse {
  id: string
  name: string
  prefix: string
  apiKey: string
  createdAt: string
}

export interface BootstrapInfo {
  proxyBaseUrl: string
  configExample: string
  hasSettings: boolean
  hasApiKey: boolean
}

// Settings API
export async function getAmpSettings(): Promise<AmpSettings> {
  const response = await authFetch(`${API_BASE}/settings`)
  return handleResponse<AmpSettings>(response)
}

export async function updateAmpSettings(data: UpdateAmpSettingsRequest): Promise<AmpSettings> {
  const response = await authFetch(`${API_BASE}/settings`, {
    method: 'PUT',
    body: JSON.stringify(data),
  })
  return handleResponse<AmpSettings>(response)
}

export async function testAmpConnection(): Promise<TestResult> {
  const response = await authFetch(`${API_BASE}/settings/test`, {
    method: 'POST',
  })
  return handleResponse<TestResult>(response)
}

// API Key
export async function getAPIKeys(): Promise<APIKey[]> {
  const response = await authFetch(`${API_BASE}/api-keys`)
  const data = await handleResponse<{ apiKeys: APIKey[] }>(response)
  return data.apiKeys || []
}

export async function createAPIKey(name: string): Promise<CreateAPIKeyResponse> {
  const response = await authFetch(`${API_BASE}/api-keys`, {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
  return handleResponse<CreateAPIKeyResponse>(response)
}

export async function deleteAPIKey(id: string): Promise<void> {
  const response = await authFetch(`${API_BASE}/api-keys/${id}`, {
    method: 'DELETE',
  })
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: '删除失败' }))
    throw new Error(error.error || '删除失败')
  }
}

export async function getAPIKey(id: string): Promise<APIKeyRevealResponse> {
  const response = await authFetch(`${API_BASE}/api-keys/${id}`)
  return handleResponse<APIKeyRevealResponse>(response)
}

// Bootstrap API
export async function getAmpBootstrap(): Promise<BootstrapInfo> {
  const response = await authFetch(`${API_BASE}/bootstrap`)
  return handleResponse<BootstrapInfo>(response)
}

// Request Logs Types
export interface RequestLog {
  id: string
  createdAt: string
  userId: string
  username?: string
  apiKeyId: string
  apiKeyName?: string
  apiKeyPrefix?: string
  originalModel?: string
  mappedModel?: string
  provider?: string
  channelName?: string
  channelId?: string
  endpoint?: string
  method: string
  path: string
  statusCode: number
  latencyMs: number
  isStreaming: boolean
  inputTokens?: number
  outputTokens?: number
  cacheReadInputTokens?: number
  cacheCreationInputTokens?: number
  errorType?: string
  requestId?: string
  costMicros?: number
  costUsd?: string
  pricingModel?: string
  thinkingLevel?: string
}

export interface RequestLogListResponse {
  items: RequestLog[]
  total: number
  page: number
  pageSize: number
}

export interface UsageSummary {
  groupKey: string
  inputTokensSum: number
  outputTokensSum: number
  cacheReadInputTokensSum: number
  cacheCreationInputTokensSum: number
  requestCount: number
  errorCount: number
  costMicrosSum?: number
  costUsdSum?: string
}

export interface UsageSummaryResponse {
  items: UsageSummary[]
}

export interface RequestLogListParams {
  page?: number
  pageSize?: number
  apiKeyId?: string
  model?: string
  status?: number
  isStreaming?: boolean
  from?: string
  to?: string
}

// Request Logs API
export async function getRequestLogs(params: RequestLogListParams = {}, signal?: AbortSignal): Promise<RequestLogListResponse> {
  const searchParams = new URLSearchParams()
  if (params.page) searchParams.set('page', params.page.toString())
  if (params.pageSize) searchParams.set('pageSize', params.pageSize.toString())
  if (params.apiKeyId) searchParams.set('apiKeyId', params.apiKeyId)
  if (params.model) searchParams.set('model', params.model)
  if (params.status !== undefined) searchParams.set('status', params.status.toString())
  if (params.isStreaming !== undefined) searchParams.set('isStreaming', params.isStreaming.toString())
  if (params.from) searchParams.set('from', params.from)
  if (params.to) searchParams.set('to', params.to)

  const query = searchParams.toString()
  const response = await authFetch(`${API_BASE}/request-logs${query ? `?${query}` : ''}`, {
    signal,
  })
  return handleResponse<RequestLogListResponse>(response)
}

export async function getUsageSummary(params: { from?: string; to?: string; groupBy?: string; model?: string } = {}, signal?: AbortSignal): Promise<UsageSummaryResponse> {
  const searchParams = new URLSearchParams()
  if (params.from) searchParams.set('from', params.from)
  if (params.to) searchParams.set('to', params.to)
  if (params.groupBy) searchParams.set('groupBy', params.groupBy)
  if (params.model) searchParams.set('model', params.model)

  const query = searchParams.toString()
  const response = await authFetch(`${API_BASE}/usage/summary${query ? `?${query}` : ''}`, {
    signal,
  })
  return handleResponse<UsageSummaryResponse>(response)
}

export async function getDistinctModels(signal?: AbortSignal): Promise<{ models: string[] }> {
  const response = await authFetch(`${API_BASE}/request-logs/models`, {
    signal,
  })
  return handleResponse<{ models: string[] }>(response)
}

// Admin API for request logs
const ADMIN_API_BASE = '/api/admin'

export interface AdminRequestLogListParams extends RequestLogListParams {
  userId?: string
}

export async function getAdminRequestLogs(params: AdminRequestLogListParams = {}, signal?: AbortSignal): Promise<RequestLogListResponse> {
  const searchParams = new URLSearchParams()
  if (params.page) searchParams.set('page', params.page.toString())
  if (params.pageSize) searchParams.set('pageSize', params.pageSize.toString())
  if (params.userId) searchParams.set('userId', params.userId)
  if (params.apiKeyId) searchParams.set('apiKeyId', params.apiKeyId)
  if (params.model) searchParams.set('model', params.model)
  if (params.status !== undefined) searchParams.set('status', params.status.toString())
  if (params.isStreaming !== undefined) searchParams.set('isStreaming', params.isStreaming.toString())
  if (params.from) searchParams.set('from', params.from)
  if (params.to) searchParams.set('to', params.to)

  const query = searchParams.toString()
  const response = await authFetch(`${ADMIN_API_BASE}/request-logs${query ? `?${query}` : ''}`, {
    signal,
  })
  return handleResponse<RequestLogListResponse>(response)
}

export async function getAdminDistinctModels(signal?: AbortSignal): Promise<{ models: string[] }> {
  const response = await authFetch(`${ADMIN_API_BASE}/request-logs/models`, {
    signal,
  })
  return handleResponse<{ models: string[] }>(response)
}

export interface DistinctAPIKey {
  id: string
  name: string
  prefix: string
}

export async function getAdminDistinctKeys(userId?: string, signal?: AbortSignal): Promise<{ keys: DistinctAPIKey[] }> {
  const searchParams = new URLSearchParams()
  if (userId) searchParams.set('userId', userId)
  const query = searchParams.toString()
  const response = await authFetch(`${ADMIN_API_BASE}/request-logs/keys${query ? `?${query}` : ''}`, {
    signal,
  })
  return handleResponse<{ keys: DistinctAPIKey[] }>(response)
}

export async function getAdminUsageSummary(params: { from?: string; to?: string; groupBy?: string; userId?: string; model?: string } = {}, signal?: AbortSignal): Promise<UsageSummaryResponse> {
  const searchParams = new URLSearchParams()
  if (params.from) searchParams.set('from', params.from)
  if (params.to) searchParams.set('to', params.to)
  if (params.groupBy) searchParams.set('groupBy', params.groupBy)
  if (params.userId) searchParams.set('userId', params.userId)
  if (params.model) searchParams.set('model', params.model)

  const query = searchParams.toString()
  const response = await authFetch(`${ADMIN_API_BASE}/usage/summary${query ? `?${query}` : ''}`, {
    signal,
  })
  return handleResponse<UsageSummaryResponse>(response)
}

// Request Log Detail API
export interface RequestLogDetail {
  requestId: string
  requestHeaders: Record<string, string>
  requestBody: string
  translatedRequestBody?: string
  responseHeaders: Record<string, string>
  responseBody: string
  translatedResponseBody?: string
  createdAt: string
}

export async function getAdminRequestLogDetail(logId: string, signal?: AbortSignal): Promise<RequestLogDetail> {
  const response = await authFetch(`${ADMIN_API_BASE}/request-logs/${logId}/detail`, {
    signal,
  })
  return handleResponse<RequestLogDetail>(response)
}
